[InstCombine] Relax one-use requirement for add iN (sext i1 X), (sext i1 Y) --> sext (X | Y) to iN

AZero13 opened the pull request:
Since these remove instructions as long as at least one of X or Y is one-use, we don't need to check one-use for both.

nikic reviewed (changes requested):
Increases instruction count if both sexts have extra uses.

AtariDreams commented:
I guess I'm gonna test that then

AtariDreams commented:
> Increases instruction count if both sexts have extra uses.

Just checked and at worst the count stays the same.

topperc commented:
> > Increases instruction count if both sexts have extra uses.
> 
> 
> 
> Just checked and at worst the count stays the same.

In the worst case doesn't it remove 1 Add and insert an Or and a SExt. That's an increase of 1.

nikic reviewed (changes requested):
The case with two extra uses is still untested?

nikic commented on llvm/test/Transforms/InstCombine/add.ll:
```diff
@@ -1435,34 +1435,52 @@ define i32 @and31_add_sexts(i1 %x, i1 %y) {
 ; CHECK-NEXT:    ret i32 [[R]]
 ;
   %xs = sext i1 %x to i32
   store i32 %xs, ptr %p
```
You already have an extra use of %xs here. Do the same for %ys.

AtariDreams replied:
I thought I did by having this use function take 2 arguments?

nikic replied:
What I meant is that you are using two different ways to introduce extra uses in the same test. You have an extra use of %xs via store here, and then below you introduce a function call to get *another* use of %xs and %ys. You should either use two stores, or drop the first store.

AZero13 committed:
[InstCombine] Pre-commit tests (NFC)

nikic commented on llvm/test/Transforms/InstCombine/add.ll:
```diff
@@ -1435,34 +1435,65 @@ define i32 @and31_add_sexts(i1 %x, i1 %y) {
+; CHECK-NEXT:    ret i32 [[R]]
+;
+  %xs = sext i1 %x to i32
+  %ys = sext i1 %y to i32
```
Now you've also dropped the extra use of %ys in this test...

AtariDreams replied:
Moved to different test.

AZero13 committed:
[InstCombine] Remove one-use requirement for add iN (sext i1 X), (sext i1 Y) --> sext (X | Y) to iN

Since these remove instructions, we can forgo the one-use check, as long as at least ONE of the two sexts are one-use.

AtariDreams commented:
@dtcxzyw Thoughts on this?

AtariDreams commented:
@nikic @topperc I believe this is ready!

AtariDreams commented:
@nikic I fixed the tests and everything!

goldsteinn commented:
The the title should the "Relax one-use..." instead of "Remove one-use..."

AtariDreams commented:
> The the title should the "Relax one-use..." instead of "Remove one-use..."

Fixed!

nikic reviewed (approved):
LGTM

goldsteinn commented:
This appears to have a pretty significant compile time impact: https://llvm-compile-time-tracker.com/compare.php?from=519e0bb094bd9444b64ac62f2e8192543fdb94b9&to=dd0245e3d74955ace14c95d288abf72346ad52e8&stat=instructions:u

nikic commented:
@goldsteinn That range also contains other commits. The regression is mostly https://github.com/llvm/llvm-project/pull/96851.
